### ISSUE TITLE
problem deploying on azure

### ISSUE THREAD

**Issue by sergiocontrino** [intermine/humanmine#2]
Error loading humanmine on remote server with postgres 9.5 and tomcat 8.5

```
12-Dec-2018 10:22:16.617 WARNING [http-nio-8080-exec-26] org.apache.catalina.core.ApplicationDispatcher.invoke Servlet [action] is currently unavailable
12-Dec-2018 10:23:26.928 INFO [http-nio-8080-exec-18] org.apache.catalina.core.ApplicationContext.log Marking servlet [action] as unavailable
12-Dec-2018 10:23:26.928 SEVERE [http-nio-8080-exec-18] org.apache.catalina.core.StandardContext.loadOnStartup Servlet [action] in web application [/humanmine] threw load() exception
 java.lang.ClassNotFoundException: org.apache.logging.log4j.util.ReflectionUtil
        at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1285)
        at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1119)
    ```

**Comment by julie-sullivan** [comment]
This is our exact problem:

https://stackoverflow.com/questions/52700803/im-getting-noclassdeffounderror-org-apache-logging-log4j-util-reflectionutil

It says the issue is because of mixed versions on the classpath. One of our dependencies must have a different version as a dep. How to find?

**Comment by sergiocontrino** [comment]
drwxrwxr-x  3 flymine 4.0K Dec 12 10:40 log4j-1.2-api/
drwxrwxr-x  3 flymine 4.0K Dec 12 10:40 log4j-api/
drwxrwxr-x  3 flymine 4.0K Dec 12 10:40 log4j-core/
drwxrwxr-x  3 flymine 4.0K Dec 12 10:40 log4j-slf4j-impl/

**Comment by sergiocontrino** [comment]
to find the dep, maybe this would help?
https://stackoverflow.com/questions/22896243/maven-slf4j-class-path-contains-multiple-slf4j-bindings/22896270

**Comment by sergiocontrino** [comment]
./gradlew -q :webapp:dependencies

[depend.txt](https://github.com/intermine/humanmine/files/2675579/depend.txt)

**Comment by sergiocontrino** [comment]
it seems we have all 2.10 version for log4j. 
there is also a org.slf4j:log4j-over-slf4j:1.7.12
in
<pre>
grettyRunnerTomcat8
\--- org.akhikhl.gretty:gretty-runner-tomcat8:2.0.0
     +--- org.akhikhl.gretty:gretty-runner-tomcat:2.0.0
     |    +--- org.akhikhl.gretty:gretty-runner:2.0.0
     |    |    +--- org.codehaus.groovy:groovy:2.4.11
     |    |    +--- org.codehaus.groovy:groovy-json:2.4.11
     |    |    |    \--- org.codehaus.groovy:groovy:2.4.11
     |    |    +--- commons-cli:commons-cli:1.2
     |    |    +--- commons-io:commons-io:2.4
     |    |    \--- ch.qos.logback:logback-classic:1.1.3
     |    |         +--- ch.qos.logback:logback-core:1.1.3
     |    |         \--- org.slf4j:slf4j-api:1.7.7 -> 1.7.12
     |    \--- org.slf4j:log4j-over-slf4j:1.7.12
     |         \--- org.slf4j:slf4j-api:1.7.12
     +--- org.apache.tomcat.embed:tomcat-embed-core:8.0.44
     +--- org.apache.tomcat.embed:tomcat-embed-el:8.0.44
     +--- org.apache.tomcat.embed:tomcat-embed-jasper:8.0.44
     |    +--- org.apache.tomcat.embed:tomcat-embed-core:8.0.44
     |    +--- org.apache.tomcat.embed:tomcat-embed-el:8.0.44
     |    \--- org.eclipse.jdt:ecj:3.12.3
     +--- org.apache.tomcat.embed:tomcat-embed-logging-log4j:8.0.44
     +--- org.apache.tomcat.embed:tomcat-embed-websocket:8.0.44
     |    \--- org.apache.tomcat.embed:tomcat-embed-core:8.0.44
     \--- javax.servlet:javax.servlet-api:3.1.0
<pre>

**Comment by sergiocontrino** [comment]
https://www.slf4j.org/legacy.html#log4j-over-slf4j

**Comment by sergiocontrino** [comment]
trying to impose the use of 2.8.1 in  webtasks 
<pre>
\--- org.intermine:intermine-webtasks:3.1.+ -> 3.1.0
     +--- ant:ant:1.6.5
     +--- org.apache.logging.log4j:log4j-1.2-api:2.10.0 (*)
     +--- org.slf4j:slf4j-simple:1.7.7 (*)
     +--- org.intermine:intermine-resources:3.1.0
     \--- org.intermine:intermine-webapp:3.1.0 (*)
</pre>

changing the build.gradle in the intermine/intermine directory
`compile group: 'org.apache.logging.log4j', name: 'log4j-1.2-api', version: '2.8.1'`

but it is automatically updated to 2.10

<pre>
\--- org.intermine:intermine-webtasks:3.1.+ -> 3.1.0
     +--- ant:ant:1.6.5
     +--- org.apache.logging.log4j:log4j-1.2-api:2.8.1 -> 2.10.0 (*)
     +--- org.slf4j:slf4j-simple:1.7.7 (*)
     +--- org.intermine:intermine-resources:3.1.0
     \--- org.intermine:intermine-webapp:3.1.0 (*)
</pre>

**Comment by sergiocontrino** [comment]
to avoid the issue:
**1)**
`rm -rf ~/.gradle/caches/modules-2/files-2/org.apache.logging.log4j/log4j-1.2-api/2.10.0/`
**2)**
in your intermine checkout:
change your build.gradle (updating the log4j-1.2-api version to an older version) to match this one
[build.gradle.txt](https://github.com/intermine/humanmine/files/2687178/build.gradle.txt)
**3)**
install intermine locally. In your intermine checkout:
```
cd intermine
./gradlew install
```
**4)**
in your mine checkout:
a)
rm this line 
`org.intermine.model.bio.MPTerm.autocomplete = name`
from the file
dbmodel/resources/objectstoresummary.config.properties
b)
```
./gradlew clean
./gradlew cargoReDeployRemote -stacktrace --no-daemon
```

checking if all of **2)** is necessary

**Comment by kkarra** [comment]
I did 1) ,  for 2) it is very different from my version of the file and am not sure which portion should match.
3a) - I don't have that line in the file 3b) works with my version of build.gradle
 but still blank page

**Comment by danielabutano** [comment]
@kkarra we have updated the instructions, please try again

**Comment by sergiocontrino** [comment]
hi @kkarra, we are using intermine release 3.1.0.

in the build.gradle of intermine:

**a)** if you have a line like
`compile group: 'org.apache.logging.log4j', name: 'log4j-1.2-api', version: '2.10.0'`
substitute it with
`compile group: 'org.apache.logging.log4j', name: 'log4j-1.2-api', version: '2.8.2'`

**b)** at the top of the file, after the 'repositories' section, add 
<pre>
    configurations {
        all{
            resolutionStrategy {
                force 'org.apache.logging.log4j:log4j-1.2-api:2.8.2'
            }
        }
    }
</pre>

**Comment by sergiocontrino** [comment]
hi @kkarra, did that work for you? thanks!

**Comment by kkarra** [comment]
Was away for the holidays..sorry for the delay.
No, it didn't work..

background - Julie had migrated stuff for me and I pulled the repos
https://github.com/julie-sullivan/yeastmine-bio-sources
https://github.com/julie-sullivan/yeastmine

So, I added your recommended changes to build.gradle in yeastmine directory and ran build commands as listed here..
https://intermine.readthedocs.io/en/latest/system-requirements/software/gradle/

While all other steps work fine.. webapp deploy is still a problem..

kalpana-thunder:yeastmine kkarra$ ./gradlew cargoDeployRemote
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=2000M; support was removed in 8.0

> Configure project :
Using properties file: '/Users/kkarra/.intermine/yeastmine.properties'
cleaning
deleting idresolver cache
deleting datasources.xml
deleting organisms.xml
deleting publications.xml
deleted log files

> Task :webapp:unwarBioWebApp
    [unzip] Expanding: /Users/kkarra/.gradle/caches/modules-2/files-2.1/org.intermine/bio-webapp/3.1.0/a17c261276bb6a779458384006242f20ce242b8e/bio-webapp-3.1.0.war into /Users/kkarra/Dev/gradle_intermine/yeastmine/webapp/build/explodedWebApp

> Task :webapp:summariseObjectStore
[summarizeObjectStore] SLF4J: Class path contains multiple SLF4J bindings.
[summarizeObjectStore] SLF4J: Found binding in [jar:file:/Users/kkarra/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-simple/1.7.7/8095d0b9f7e0a9cd79a663c740e0f8fb31d0e2c8/slf4j-simple-1.7.7.jar!/org/slf4j/impl/StaticLoggerBinder.class]
[summarizeObjectStore] SLF4J: Found binding in [jar:file:/Users/kkarra/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-jdk14/1.7.10/13675a57601d9106fe82314fe9eebb9b2d9b13c6/slf4j-jdk14-1.7.10.jar!/org/slf4j/impl/StaticLoggerBinder.class]
[summarizeObjectStore] SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
[summarizeObjectStore] SLF4J: Actual binding is of type [org.slf4j.impl.SimpleLoggerFactory]
[summarizeObjectStore] [main] INFO com.zaxxer.hikari.HikariDataSource - db.production - Starting...
[summarizeObjectStore] [main] INFO com.zaxxer.hikari.HikariDataSource - db.production - Start completed.


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':webapp:cargoDeployRemote'.
> org.codehaus.cargo.container.ContainerException: Failed to deploy [/Users/kkarra/Dev/gradle_intermine/yeastmine/webapp/build/libs/webapp.war]

Not sure it matters but the re-deploy commands works, but I still get a blank page. 

kalpana-thunder:yeastmine kkarra$ ./gradlew cargoRedeployRemote --stacktrace --no-daemon
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=2000M; support was removed in 8.0

> Configure project :
Using properties file: '/Users/kkarra/.intermine/yeastmine.properties'
cleaning
deleting idresolver cache
deleting datasources.xml
deleting organisms.xml
deleting publications.xml
deleted log files

> Task :dbmodel:createSoModel
[createSoModel] Total terms: 2219
[createSoModel] Total terms, post-pruning: 73
[createSoModel] Total terms, post-flattening: 42
[createSoModel] Wrote /Users/kkarra/Dev/gradle_intermine/yeastmine/dbmodel/build/so_additions.xml

> Task :dbmodel:mergeModels
[mergeSourceModels] warning: sgd-gff_additions.xml not found
[mergeSourceModels] warning: update-publications_additions.xml not found
[mergeSourceModels] warning: entrez-organism_additions.xml not found
[mergeSourceModels] log4j:WARN No appenders could be found for logger (org.intermine.modelproduction.ModelMerger).
[mergeSourceModels] log4j:WARN Please initialize the log4j system properly.
[mergeSourceModels] log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.

> Task :dbmodel:compileJava
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

> Task :webapp:unwarBioWebApp
    [unzip] Expanding: /Users/kkarra/.gradle/caches/modules-2/files-2.1/org.intermine/bio-webapp/3.1.0/a17c261276bb6a779458384006242f20ce242b8e/bio-webapp-3.1.0.war into /Users/kkarra/Dev/gradle_intermine/yeastmine/webapp/build/explodedWebApp

> Task :webapp:summariseObjectStore
[summarizeObjectStore] SLF4J: Class path contains multiple SLF4J bindings.
[summarizeObjectStore] SLF4J: Found binding in [jar:file:/Users/kkarra/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-simple/1.7.7/8095d0b9f7e0a9cd79a663c740e0f8fb31d0e2c8/slf4j-simple-1.7.7.jar!/org/slf4j/impl/StaticLoggerBinder.class]
[summarizeObjectStore] SLF4J: Found binding in [jar:file:/Users/kkarra/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-jdk14/1.7.10/13675a57601d9106fe82314fe9eebb9b2d9b13c6/slf4j-jdk14-1.7.10.jar!/org/slf4j/impl/StaticLoggerBinder.class]
[summarizeObjectStore] SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
[summarizeObjectStore] SLF4J: Actual binding is of type [org.slf4j.impl.SimpleLoggerFactory]
[summarizeObjectStore] [Task worker for ':'] INFO com.zaxxer.hikari.HikariDataSource - db.production - Starting...
[summarizeObjectStore] [Task worker for ':'] INFO com.zaxxer.hikari.HikariDataSource - db.production - Start completed.


BUILD SUCCESSFUL in 52s
22 actionable tasks: 22 executed
[Thread-17] INFO com.zaxxer.hikari.HikariDataSource - db.production - Shutdown initiated...
[Thread-17] INFO com.zaxxer.hikari.HikariDataSource - db.production - Shutdown completed.


contents of build.gradle

buildscript {

    configurations {
        all{
            resolutionStrategy {
                force 'org.apache.logging.log4j:log4j-1.2-api:2.8.2'
            }
        }
    }

    repositories {
        mavenLocal()
        jcenter()
        maven {
            url "https://oss.jfrog.org/artifactory/oss-snapshot-local"
        }
    }
    dependencies {
        classpath group: 'org.intermine', name: 'plugin', version: System.getProperty("imVersion")
        classpath 'com.bmuschko:gradle-cargo-plugin:2.3'
    }
}

ext {
  mineRelease = project.hasProperty('release') ? "." + release : ''
  mineName = "yeastmine"
  minePropertyFileName = "${mineName}.properties$mineRelease"
  minePropertyFile = "${System.env.HOME}/.intermine/$minePropertyFileName"
  println "Using properties file: '$minePropertyFile'"
}

subprojects {
  apply plugin: 'java'
  apply plugin: 'maven'

  sourceCompatibility = 1.8
  targetCompatibility = 1.8
  
  repositories {
    mavenLocal()
    jcenter()
    maven { url 'https://clojars.org/repo' }
    maven {
        url "https://oss.jfrog.org/artifactory/oss-snapshot-local"
    }
    maven {
        url 'http://www.ebi.ac.uk/intact/maven/nexus/content/repositories/ebi-repo/'
    }
  }

  dependencies {
    compile group: "org.intermine", name: "intermine-resources", version: System.getProperty("imVersion")
    compile('ant:ant:1.6.5')
    compile group: 'org.apache.logging.log4j', name: 'log4j-1.2-api', version: '2.8.2'
    testCompile group: 'junit', name: 'junit', version: '4.8.2'
  }

  task copyMineProperties(type: Copy) {
    description "Copies mine specific intermine.properties file (from .intermine directory) into resources output to be included in the war"
    dependsOn 'processResources'
    from (minePropertyFile)
    into sourceSets.main.output.resourcesDir
    rename { fileName -> fileName.replace("$minePropertyFileName", "intermine.properties") }
    inputs.sourceFiles.stopExecutionIfEmpty()
  }
}

task clean(type: Delete) {
    println "cleaning"
    delete "idresolver.cache"
    println "deleting idresolver cache"
    delete "datasources.xml"
    println "deleting datasources.xml"
    delete "organisms.xml"
    println "deleting organisms.xml"
    delete "publications.xml"
    println "deleting publications.xml"
    delete fileTree('.') {
        include '*.log'
    }
    println "deleted log files"
}

**Comment by kkarra** [comment]
I will try to check discord tomorrow 8am pst 4pm uk time

**Comment by sergiocontrino** [comment]
hi @kkarra , you should do the change in your local intermine (3.1.0) directory (rather than the yeastmine one). apologies if this was not clear. i will make a test branch that you can clone directly (either today or tomorrow..).

**Comment by sergiocontrino** [comment]
hi @kkarra , you can use the branch intermine-KK in https://github.com/sergiocontrino/intermine
which has the corrected intermine/build.gradle file (but i realise now that probably the differences you were seeing before were due to the wrong repository, so probably you can just substitute the file i sent then https://github.com/intermine/humanmine/files/2687178/build.gradle.txt)
please let me know if it works (please do again also steps 1,3 and 4)

**Comment by kkarra** [comment]
I don't understand.  There is no intermine folder where I am trying this.. just yeastmine/ and yeastmine-bio-sources/  folder. I thought we don't have to have intermine/ locally.

**Comment by sergiocontrino** [comment]
hi @kkarra, the error you are getting is not happening always, so it could be interesting for us if  you do this test. possibly unconvenient, so in alternative we could release a new version of intermine (this will follow the usual release process but should be reasonably quick). i'll check with @julie-sullivan and update later.

**Comment by sergiocontrino** [comment]
hi @kkarra, please try the local install and the fix proposed. we are not considering a release just now, and the fix could create issues for instances that are working with the current release.

**Comment by kkarra** [comment]
My bad..sorry..

I interpreted "install intermine locally. In your intermine checkout:..." differently...
All I had heard/read seemed to indicate you don't need intermine locally, so, I don't know what I did instead :) - I should have asked for clarification.

So, now that I cloned intermine-KK branch and ran step 3) correctly. I am able to go through the webapp deploy step and the home page shows up correctly!! 

Thanks for your patience!!
